ilastik/applets/multicut/opMulticut.py

```diff
@@ -1,518 +1,306 @@
-        probabilities = np.random.random(superpixels.shape[:-1] + (3,)).astype(np.float32)
-        probabilities = vigra.taggedView(probabilities, "zyxc")
+        Returns: An index array [0,1,...,N] indicating the new labels for the N nodes of the RAG.
+        """
+        #
+        # Check parameters
+        #
+        assert rag.edge_ids.shape == (rag.num_edges, 2)
+        node_count = rag.max_sp + 1
+        #
+        # Solve
+        #
+        edge_weights = compute_edge_weights(rag.edge_ids, edge_probabilities, beta, threshold)
+        assert edge_weights.shape == (rag.num_edges,)
+
+        mapping_index_array = solve(rag.edge_ids, edge_weights, node_count, solver_name)
 
-        superpixels = vigra.taggedView(superpixels, "zyxc")
+        return mapping_index_array
 
-        from lazyflow.graph import Graph
 
-        op = OpMulticut(graph=Graph())
-        op.VoxelData.setValue(probabilities)
-        op.InputSuperpixels.setValue(superpixels)
-        assert op.Output.ready()
-        seg = op.Output[:].wait()
+def compute_edge_weights(edge_ids, edge_probabilities, beta, threshold):
+    """
+    Convert edge probabilities to energies for the multicut problem.
 
-        assert seg.min() == 0
+    edge_ids:
+        The list of edges in the graph. shape=(N, 2)
+    edge_probabilities:
+        1-D, float (1.0 means edge is CUT, disconnecting the two SPs)
+    beta:
+        scalar (float)
+    threshold:
+        scalar (float), moves the 0 of the edge weights (default threshold = 0.5)
 
-        print("DONE.")
+    Special behavior:
+        If any node has ID 0, all of it's edges will be given an
+        artificially low energy, to prevent it from merging with its
+        neighbors, regardless of what the edge_probabilities say.
+    """
+
+    def rescale(probabilities, threshold):
+        """
+        Given a threshold in the range (0,1), rescales the probabilities below and above
+        the threshold to the ranges (0,0.5], and (0.5,1) respectively. This is needed
+        to implement an effective 'moving' of the 0 weight, since the multicut algorithm
+        implicitly calculates that weights change sign at p=0.5.
+
+        :param probabilities: 1d array (float). Probability data within range (0,1)
+        :param threshold: scalar (float). The new threshold for the algorithm.
+        :return: Rescaled data to be used in algorithm.
+        """
+        out = np.zeros_like(probabilities)
+        data_lower = probabilities[probabilities <= threshold]
+        data_upper = probabilities[probabilities > threshold]
+
+        data_lower = (data_lower / threshold) * 0.5
+        data_upper = (((data_upper - threshold) / (1 - threshold)) * 0.5) + 0.5
+
+        out[probabilities <= threshold] = data_lower
+        out[probabilities > threshold] = data_upper
+        return out
+
+    p1 = edge_probabilities  # P(Edge=CUT)
+    p1 = np.clip(p1, 0.001, 0.999)
+    p1 = rescale(p1, threshold)
```

Comment:
```suggestion
    # Rescale 0; t] to [0; 0.5], and [t; 1] to [0.5; 1].
    p1 = np.where(p1 <= threshold, p1 / (2 * threshold), 0.5 + (p1 - threshold) / (2 * (1 - threshold)))
```

Reply:
you're version is not only easier to read, it's also 4x faster :)